src/GL/Program.hs

```diff
@@ -78,36 +75,27 @@ build p = runLiftIO $ do
   pure (Program ls program)
 
 use :: Has (Lift IO) sig m => Program u v o -> ProgramC u v o m a -> m a
-use (Program ls p) (ProgramC m) = do
+use (Program ls p) m = do
   sendIO (glUseProgram p)
-  runReader (Program ls p) m
+  runReader (Program ls p) (runProgramC m)
 
 
-class Monad m => HasProgram (u :: (* -> *) -> *) (v :: (* -> *) -> *) (o :: (* -> *) -> *) (m :: * -> *) | m -> u v o where
-  askProgram :: m (Program u v o)
+type HasProgram u v o sig m = HasLabelled Program (Reader (Program u v o)) sig m
 
+askProgram :: HasLabelled Program (Reader (Program u v o)) sig m => m (Program u v o)
+askProgram = runUnderLabel @_ @Program ask
 
-newtype ProgramC (u :: (* -> *) -> *) (v :: (* -> *) -> *) (o :: (* -> *) -> *) m a = ProgramC { runProgramT :: ReaderC (Program u v o) m a }
-  deriving (Applicative, Functor, Monad, MonadIO, MonadTrans)
-
-instance HasProgram u v o m => HasProgram u v o (ReaderC r m) where
-  askProgram = lift askProgram
 
-instance HasProgram u v o m => HasProgram u v o (StateC s m) where
-  askProgram = lift askProgram
+newtype ProgramC (u :: (* -> *) -> *) (v :: (* -> *) -> *) (o :: (* -> *) -> *) m a = ProgramC { runProgramC :: ReaderC (Program u v o) m a }
+  deriving (Applicative, Functor, Monad, MonadIO, MonadTrans)
 
-instance (Has Check sig m, Has (Lift IO) sig m, Vars u) => Algebra (State (u Maybe) :+: sig) (ProgramC u v o m) where
+instance (Has Check sig m, Has (Lift IO) sig m, Vars u) => Algebra (State (u Maybe) :+: Labelled Program (Reader (Program u v o)) :+: sig) (ProgramC u v o m) where
   alg = \case
     L (Get   k) -> k (makeVars (const Nothing))
     L (Put s k) -> do
       Program ls prog <- askProgram
       foldVarsM (\ Field { location, value } ->
         maybe (pure ()) (checking . uniform prog (ls IntMap.! location)) value) s
       k
-    R other     -> ProgramC (send (handleCoercible other))
-
-instance Algebra sig m => HasProgram u v o (ProgramC u v o m) where
-  askProgram = ProgramC ask
-
-deriving instance B.HasBuffer ty x m => B.HasBuffer ty x (ProgramC u v o m)
-deriving instance HasProgram u v o m => HasProgram u v o (B.BufferC ty x m)
+    R (L other) -> ProgramC (send (handleCoercible (runLabelled other)))
+    R (R other) -> ProgramC (send (handleCoercible other))
```

Comment:
Here we’ve opted to retain our existing `ProgramC` carrier and enrich it with handling the `Labelled` effect directly; this avoids the need for other modules (e.g. `UI.Drawable`) to care about `Labelled` at all.

This serves to reinforce the fact that there is nothing particularly special about `Labelled` or even about labelled effects; a fact which is already evident from it happening here, in a game being written in another repo much later on with very different goals in mind. It further illustrates that functional dependencies are not the exclusive province of `mtl`; `fused-effects` allows you to employ them precisely when you wish to.